File: leetcode-submissions/678_Valid_Parenthesis_String.cpp
class Solution {
public:
    bool checkValidString(string s) {
        // now i am writing the solution of this question by using the stack approach in which i am using the two stacks one is open and second one is star
        stack<int>open;
        stack<int>star;
        for(int i=0;i<s.size();i++){
            //if we find the opening bracket then push into the open stack
            if(s[i]=='('){
                open.push(i);

            }else if(s[i]=='*'){
                star.push(i);

            }else if(not open.empty() and s[open.top()]=='(' and s[i]==')'){
                open.pop();

            }
        }
        //now writing the code for start case now we are travbersing full of the open stack 
        while(not open.empty()){
            if(star.empty()){
                return false;
            }
            else if(open.top()<star.top()){
                open.pop();
                star.pop();

            }else{
                return false;
            }
        }
        return true;
        
    }
};

File: leetcode-submissions/678_Valid_Parenthesis_String.ts
function checkValidString(s: string): boolean {
    //now i am solving this question by using the array as stack 
    const open:number[]=[];
    const star:number[]=[];
    for(let i=0;i<s.length;i++){
        if(s.charAt(i)==="("){
            open.push(i);
        }
        else if(s.charAt(i)==="*"){
            star.push(i);
        }else{
if(open.length>0){
    open.pop();

}else if(star.length>0){
    star.pop();
}else{
    return false;
}
        }
    }
    while(open.length>0){
        if(star.length===0){
            return false;
        }else if(open[open.length-1]<star[star.length-1]){
            open.pop();
            star.pop();
        }else{
            return false;
        }
    }
return true ;
};